### PR TITLE
[GEP-19] Adapt monitoring configuration for shoot cluster system components

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:
As a follow-up of https://github.com/gardener/gardener/pull/9695, this PR adapts the monitoring configuration for the system components running inside the shoot cluster.

The shoot cluster control plane components will be adapted in a subsequent PR.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/9065

**Special notes for your reviewer**:
/cc @ScheererJ 

**Release note**:

```feature developer
Extensions deploying shoot cluster system components can now make use of the `pkg/component/observability/monitoring/prometheus/shoot.ClusterComponentScrapeConfigSpec` function in order to generate a [`ScrapeConfig`](https://github.com/gardener/gardener/blob/master/docs/extensions/logging-and-monitoring.md#scrapeconfigs) for the shoot's Prometheus.
```

### CODE REVIEW

File: pkg/component/observability/monitoring/prometheus/shoot/utils.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+			SourceLabels: []monitoringv1.LabelName{"__meta_kubernetes_service_name", "__meta_kubernetes_endpoint_port_name"},
+			Action:       "keep",
+			Regex:        sdConfig.ServiceName + `;` + sdConfig.EndpointPortName,
+		}}
+	}
+
+	return monitoringv1alpha1.ScrapeConfigSpec{
+		HonorLabels: ptr.To(false),
+		Scheme:      ptr.To("HTTPS"),
+		// This is needed because the kubelets' certificates are not are generated for a specific pod IP
```
Comment (ScheererJ) [inline pkg/component/observability/monitoring/prometheus/shoot/utils.go]: Are the metrics really retrieved via kubelet and not via direct connection to the pod? If `kubelet` is always in the loop the comment makes sense.
I thought, though, that prometheus/blackbox-exporter scrape the pods via `kube-apiserver`, but without going through `kubelet`. Then the comment would be misleading.

WDYT?

File: pkg/component/observability/monitoring/prometheus/shoot/utils.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+		KubernetesSDConfigs: []monitoringv1alpha1.KubernetesSDConfig{{
+			Role:       sdConfig.Role,
+			APIServer:  ptr.To("https://" + v1beta1constants.DeploymentNameKubeAPIServer),
+			Namespaces: &monitoringv1alpha1.NamespaceDiscovery{Names: []string{metav1.NamespaceSystem}},
+			Authorization: &monitoringv1.SafeAuthorization{Credentials: &corev1.SecretKeySelector{
+				LocalObjectReference: corev1.LocalObjectReference{Name: AccessSecretName},
+				Key:                  resourcesv1alpha1.DataKeyToken,
+			}},
+			// This is needed because we do not fetch the correct cluster CA bundle right now
+			TLSConfig: &monitoringv1.SafeTLSConfig{InsecureSkipVerify: true},
```
Comment (ScheererJ) [inline pkg/component/observability/monitoring/prometheus/shoot/utils.go]: If this relates to `kube-apiserver`'s CA bundle it should be possible to provide it to prometheus/blackbox-exporter. Is there any particular reason why this was not done or are there any complications I naively overlooked? From my point of view, this seems to be easier to 'fix' than the `kubelet` CA problem as the cluster CA should be available in the control plane.

Comment (ScheererJ) [reply pkg/component/observability/monitoring/prometheus/shoot/utils.go]: Unfortunately, the [documentation](https://kubernetes.io/docs/tasks/access-application-cluster/access-cluster-services/#discovering-builtin-services) does not mention whether `kubelet` is in the loop or not.

File: pkg/component/networking/nodelocaldns/nodelocaldns_test.go [hunk 3]
```diff
@@ -914,15 +1082,19 @@ ip6.arpa:53 {
...
 		It("should successfully destroy all resources", func() {
+			scrapeConfig.ResourceVersion = ""
+			scrapeConfigErrors.ResourceVersion = ""
+
 			component = New(c, namespace, values)
 			Expect(c.Create(ctx, managedResource)).To(Succeed())
 			Expect(c.Create(ctx, managedResourceSecret)).To(Succeed())
-
-			Expect(c.Get(ctx, client.ObjectKeyFromObject(managedResource), managedResource)).To(Succeed())
-			Expect(c.Get(ctx, client.ObjectKeyFromObject(managedResourceSecret), managedResourceSecret)).To(Succeed())
```
Comment (ScheererJ) [inline pkg/component/networking/nodelocaldns/nodelocaldns_test.go:922]: Any particular reason why you removed those lines? Checking that the managed resource is available in terms of getting it seems like a valid approach.

File: pkg/component/networking/apiserverproxy/apiserver_proxy.go [hunk 4]
```diff
@@ -105,6 +107,54 @@ func (a *apiserverProxy) Deploy(ctx context.Context) error {
...
+				SourceLabels: []monitoringv1.LabelName{"envoy_cluster_name"},
+				Regex:        `^uds_admin$`,
+				Action:       "drop",
+			},
+			&monitoringv1.RelabelConfig{
+				SourceLabels: []monitoringv1.LabelName{"envoy_listener_address"},
+				Regex:        `^^0.0.0.0_16910$`,
+				Action:       "drop",
+			},
+		)
```
Comment (ScheererJ) [inline pkg/component/networking/apiserverproxy/apiserver_proxy.go]: What does this do compared to the list above? I understood the list above that we only take over the listed metrics. Now, we explicitly drop some. It seems like similar functionality.

@istvanballok / @rickardsjp : Could you comment please unless @rfranzke knows?

File: pkg/component/networking/apiserverproxy/apiserver_proxy.go [hunk 5]
```diff
@@ -33,15 +38,13 @@ import (
 )
 
 const (
-	managedResourceName   = "shoot-core-apiserver-proxy"
-	configMapName         = "apiserver-proxy-config"
-	serviceAcountName     = "apiserver-proxy"
-	daemonSetName         = "apiserver-proxy"
-	mutatingWebhookName   = "apiserver-proxy.networking.gardener.cloud"
-	webhookExpressionsKey = "apiserver-proxy.networking.gardener.cloud/inject"
```
Comment (ScheererJ) [inline pkg/component/networking/apiserverproxy/apiserver_proxy.go:41]: At least in this commit, it looks like these constants were never used. How come that the `go` compiler does not complain about them?

File: pkg/component/observability/monitoring/nodeexporter/nodeexporter.go [hunk 6]
```diff
@@ -81,15 +78,184 @@ type nodeExporter struct {
...
+						Expr:  intstr.FromString(`instance:conntrack_entries_usage:percent > 90`),
+						For:   ptr.To(monitoringv1.Duration("1h")),
+						Labels: map[string]string{
+							"service":    name,
+							"severity":   "critical",
+							"type":       "shoot",
+							"visibility": "owner",
+						},
+						Annotations: map[string]string{
+							"description": "The nf_conntrack table is{{$value}}% full.",
```
Comment (ScheererJ) [inline pkg/component/observability/monitoring/nodeexporter/nodeexporter.go]:
```suggestion
							"description": "The nf_conntrack table is {{$value}}% full.",
```

File: pkg/component/observability/monitoring/nodeexporter/nodeexporter.go [hunk 7]
```diff
@@ -81,15 +78,184 @@ type nodeExporter struct {
...
+						For:   ptr.To(monitoringv1.Duration("1h")),
+						Labels: map[string]string{
+							"service":    name,
+							"severity":   "warning",
+							"type":       "shoot",
+							"visibility": "owner",
+						},
+						Annotations: map[string]string{
+							"summary":     "Node is under disk pressure.",
+							"description": "Node{{$labels.node}} is under disk pressure.",
```
Comment (ScheererJ) [inline pkg/component/observability/monitoring/nodeexporter/nodeexporter.go]:
```suggestion
							"description": "Node {{$labels.node}} is under disk pressure.",
```

File: pkg/component/observability/monitoring/nodeexporter/nodeexporter.go [hunk 8]
```diff
@@ -81,15 +78,184 @@ type nodeExporter struct {
...
+						For:   ptr.To(monitoringv1.Duration("1h")),
+						Labels: map[string]string{
+							"service":    name,
+							"severity":   "warning",
+							"type":       "shoot",
+							"visibility": "owner",
+						},
+						Annotations: map[string]string{
+							"summary":     "Node is under memory pressure.",
+							"description": "Node{{$labels.node}} is under memory pressure.",
```
Comment (ScheererJ) [inline pkg/component/observability/monitoring/nodeexporter/nodeexporter.go]:
```suggestion
							"description": "Node {{$labels.node}} is under memory pressure.",
```

File: pkg/component/observability/monitoring/nodeexporter/nodeexporter.go [hunk 9]
```diff
@@ -81,15 +78,184 @@ type nodeExporter struct {
...
+						For:   ptr.To(monitoringv1.Duration("1h")),
+						Labels: map[string]string{
+							"service":    name,
+							"severity":   "critical",
+							"type":       "shoot",
+							"visibility": "owner",
+						},
+						Annotations: map[string]string{
+							"summary":     "Node ran out of disk space.",
+							"description": "Node{{$labels.node}} has run out of disk space.",
```
Comment (ScheererJ) [inline pkg/component/observability/monitoring/nodeexporter/nodeexporter.go]:
```suggestion
							"description": "Node {{$labels.node}} has run out of disk space.",
```

File: pkg/component/networking/vpn/shoot/shoot.go [hunk 10]
```diff
@@ -125,6 +126,127 @@ type vpnSecret struct {
 }
 
 func (v *vpnShoot) Deploy(ctx context.Context) error {
+	scrapeConfig := v.emptyScrapeConfig()
+	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, v.client, scrapeConfig, func() error {
+		metav1.SetMetaDataLabel(&scrapeConfig.ObjectMeta, "prometheus", shoot.Label)
+		scrapeConfig.Spec = monitoringv1alpha1.ScrapeConfigSpec{
```
Comment (ScheererJ) [inline pkg/component/networking/vpn/shoot/shoot.go:132]: Why do we not use `shoot.ClusterComponentScrapeConfigSpec(...)` here? Is this `ScrapeConfig` just too different?

File: pkg/component/networking/coredns/coredns.go [hunk 11]
```diff
@@ -119,16 +120,84 @@ type coreDNS struct {
...
+						Expr:  intstr.FromString(`absent(up{job="coredns"} == 1)`),
+						For:   ptr.To(monitoringv1.Duration("20m")),
+						Labels: map[string]string{
+							"service":    serviceName,
+							"severity":   "critical",
+							"type":       "shoot",
+							"visibility": "all",
+						},
+						Annotations: map[string]string{
+							"description": "CoreDNS could not be found.Cluster DNS resolution will not work.",
```
Comment (ScheererJ) [inline pkg/component/networking/coredns/coredns.go]:
```suggestion
							"description": "CoreDNS could not be found. Cluster DNS resolution will not work.",
```

**Review by ScheererJ — COMMENTED** [review]
Thanks a lot for simplifying the monitoring configuration further. I have a few questions/remarks.

File: pkg/component/networking/apiserverproxy/apiserver_proxy.go (hunk 5 above)
Comment (rfranzke) [reply pkg/component/networking/apiserverproxy/apiserver_proxy.go:41]: I think they were unused, yes. Go compiler only complains about unused `var`s, not about unused `const`s.

File: pkg/component/networking/apiserverproxy/apiserver_proxy.go (hunk 4 above)
Comment (rfranzke) [reply pkg/component/networking/apiserverproxy/apiserver_proxy.go]: I just reworked the existing scrape config, I have no idea about the background/motivation for such specialities. I would also prefer to move this discussion to a follow-up PR and focus on the plain refactoring here.

File: pkg/component/networking/nodelocaldns/nodelocaldns_test.go (hunk 3 above)
Comment (rfranzke) [reply pkg/component/networking/nodelocaldns/nodelocaldns_test.go:922]: IMO, this assertion is useless. We create these two objects in the two previous lines, and we assert that this succeeded. Checking whether they indeed exist is useless unless we would like to test the functionality of the fake client.

File: pkg/component/networking/vpn/shoot/shoot.go (hunk 10 above)
Comment (rfranzke) [reply pkg/component/networking/vpn/shoot/shoot.go:132]: This is a config for `blackbox-exporter`, so somewhat "special". It indeed has some differences compared to the result of `shoot.ClusterComponentScrapeConfigSpec`, so I didn't use it.

File: pkg/component/observability/monitoring/prometheus/shoot/utils.go (hunk 2 above)
Comment (rfranzke) [reply pkg/component/observability/monitoring/prometheus/shoot/utils.go]: Yes, we can fix it, but I would prefer if the colleagues @istvanballok @rickardsjp @vicwicker work on this separately after we have fully completed GEP-19. For now, I'd just like to translate what we have, and for the vast majority of the scrape configs, we already skip TLS verification as of today.

File: pkg/component/networking/nodelocaldns/nodelocaldns_test.go (hunk 3 above)
Comment (ScheererJ) [reply pkg/component/networking/nodelocaldns/nodelocaldns_test.go:922]: Ok, makes sense.

File: pkg/component/networking/apiserverproxy/apiserver_proxy.go (hunk 4 above)
Comment (ScheererJ) [reply pkg/component/networking/apiserverproxy/apiserver_proxy.go]: This is certainly not blocking in any way. It just felt like we could remove some code as we drop some things anyway, but the monitoring colleagues can also take care later on.

File: pkg/component/networking/apiserverproxy/apiserver_proxy.go (hunk 5 above)
Comment (ScheererJ) [reply pkg/component/networking/apiserverproxy/apiserver_proxy.go:41]: Then it was a good catch :-) .

**Review by ScheererJ — COMMENTED** [review]
/lgtm

**Pull request merged**